src/main/java/org/jenkinsci/deprecatedusage/Downloader.java

```diff
@@ -98,33 +108,53 @@ public void run() {
 
         private void doRun() throws IOException, DigestException {
             URL url = new URL(file.getUrl());
-            HttpURLConnection request = (HttpURLConnection) url.openConnection();
-            int responseCode = request.getResponseCode();
-            if (responseCode == 502 && retriesRemaining.getAndDecrement() > 0) {
-                // flaky update center
-                doRun();
-            } else if (responseCode >= 400) {
-                throw new HttpResponseException(responseCode, request.getResponseMessage());
-            } else {
-                long fileSize;
-                try (InputStream in = request.getInputStream();
-                     OutputStream out = file.getFileOutputStream()) {
-                    fileSize = IOUtils.copyLarge(in, out);
-                } catch (IOException e) {
-                    if ("Premature EOF".equals(e.getMessage()) && retriesRemaining.getAndDecrement() > 0) {
-                        // flaky update center
-                        doRun();
-                        return;
+            try {
+                HttpURLConnection request = (HttpURLConnection) url.openConnection();
+                int responseCode = request.getResponseCode();
+                if (responseCode == 502) {
+                    throw new IOException("Flaky Update Center returned HTTP 502");
+                } else if (responseCode >= 400) {
+                    throw new HttpResponseException(responseCode, request.getResponseMessage());
+                } else {
+                    long fileSize;
+                    try (InputStream in = request.getInputStream();
+                         OutputStream out = file.getFileOutputStream()) {
+                        fileSize = IOUtils.copyLarge(in, out);
+                    }
+                    if (file.isFileMessageDigestValid()) {
+                        System.out.printf("Downloaded %s @ %.2f kiB%n", file.getUrl(), (fileSize / 1024.0));
                     } else {
-                        throw e;
+                        throw new DigestException("Downloaded file message digest does not match update center for " + url);
                     }
                 }
-                if (file.isFileMessageDigestValid()) {
-                    System.out.printf("Downloaded %s @ %.2f kiB%n", file.getUrl(), (fileSize / 1024.0));
+            } catch (IOException ioEx) {
+                if (shouldRetryForException(ioEx) && retriesRemaining.getAndDecrement() > 0) {
+                    try {
+                        System.out.printf("Failed to download %s due to %s, will retry in 750ms%n", StringUtils.isEmpty(ioEx.getMessage()) ? ioEx.getClass().getName() : ioEx.getMessage() , file.getUrl());
+                        Thread.sleep(7500L);
```

Comment:
the old code also immediately retried - which now we handle a socket connect timeout is not the best thing.  it something prematurly closed the connection it is also generally a good idea to have a bit of a backoff

Reply:
feels like we should just move this to a modern http client?